File: test/language/explicit_parameters/java/explicit_parameters/explicit_simple_param/ExplicitSimpleParamTest.java
package explicit_parameters.explicit_simple_param;

import static org.junit.Assert.assertEquals;

import java.io.File;
import java.io.IOException;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import explicit_parameters.ExplicitParametersDb;

import zserio.runtime.ZserioError;

public class ExplicitSimpleParamTest
{
    @BeforeClass
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @Before
    public void setUp() throws IOException, SQLException
    {
        FileUtil.deleteFileIfExists(file);
        database = new ExplicitParametersDb(file.toString());
        database.createSchema();
    }

    @After
    public void tearDown() throws SQLException
    {
        if (database != null)
        {
            database.close();
            database = null;
        }
    }

    @Test
    public void readWithoutCondition() throws SQLException, IOException, ZserioError
    {
        final SimpleParamTable simpleParamTable = database.getSimpleParamTable();

        final List<SimpleParamTableRow> writtenRows = new ArrayList<SimpleParamTableRow>();
        fillSimpleParamTableRows(writtenRows);
        simpleParamTable.write(writtenRows);

        final SimpleParamTableParameterProvider parameterProvider = new SimpleParamTableParameterProvider();
        final List<SimpleParamTableRow> readRows = simpleParamTable.read(parameterProvider);
        checkSimpleParamTableRows(writtenRows, readRows);
    }

    @Test
    public void readWithCondition() throws SQLException, IOException, ZserioError
    {
        final SimpleParamTable simpleParamTable = database.getSimpleParamTable();

        final List<SimpleParamTableRow> writtenRows = new ArrayList<SimpleParamTableRow>();
        fillSimpleParamTableRows(writtenRows);
        simpleParamTable.write(writtenRows);

        final SimpleParamTableParameterProvider parameterProvider = new SimpleParamTableParameterProvider();
        final String condition = "name='Name1'";
        final List<SimpleParamTableRow> readRows = simpleParamTable.read(parameterProvider, condition);
        assertEquals(1, readRows.size());

        final int expectedRowNum = 1;
        final SimpleParamTableRow readRow = readRows.get(0);
        checkSimpleParamTableRow(writtenRows.get(expectedRowNum), readRow);
    }

    @Test
    public void update() throws SQLException, IOException, ZserioError
    {
        final SimpleParamTable simpleParamTable = database.getSimpleParamTable();

        final List<SimpleParamTableRow> writtenRows = new ArrayList<SimpleParamTableRow>();
        fillSimpleParamTableRows(writtenRows);
        simpleParamTable.write(writtenRows);

        final int updateRowId = 3;
        final SimpleParamTableRow updateRow = createSimpleParamTableRow(updateRowId, "UpdatedName");
        final String updateCondition = "id=" + updateRowId;
        simpleParamTable.update(updateRow, updateCondition);

        final SimpleParamTableParameterProvider parameterProvider = new SimpleParamTableParameterProvider();
        final List<SimpleParamTableRow> readRows = simpleParamTable.read(parameterProvider, updateCondition);
        assertEquals(1, readRows.size());

        final SimpleParamTableRow readRow = readRows.get(0);
        checkSimpleParamTableRow(updateRow, readRow);
    }

    private static class SimpleParamTableParameterProvider implements SimpleParamTable.ParameterProvider
    {
        @Override
        public long getCount1(ResultSet resultSet)
        {
            return SIMPLE_PARAM_TABLE_COUNT1;
        }

        @Override
        public long getCount2(ResultSet resultSet)
        {
            return SIMPLE_PARAM_TABLE_COUNT2;
        }
    }

    private static void fillSimpleParamTableRows(List<SimpleParamTableRow> rows)
    {
        for (int id = 0; id < NUM_SIMPLE_PARAM_TABLE_ROWS; ++id)
        {
            rows.add(createSimpleParamTableRow(id, "Name" + id));
        }
    }

    private static SimpleParamTableRow createSimpleParamTableRow(long id, String name)
    {
        final SimpleParamTableRow row = new SimpleParamTableRow();

        row.setId(id);
        row.setName(name);

        final byte[] values1 = new byte[SIMPLE_PARAM_TABLE_COUNT1];
        for (int i = 0; i < values1.length; ++i)
            values1[i] = (byte)id;
        final TestBlob testBlob1 = new TestBlob(values1.length, values1);
        row.setBlob1(testBlob1);

        final byte[] values2 = new byte[SIMPLE_PARAM_TABLE_COUNT2];
        for (int i = 0; i < values2.length; ++i)
            values2[i] = (byte)(id + 1);
        final TestBlob testBlob2 = new TestBlob(values2.length, values2);
        row.setBlob2(testBlob2);

        final byte[] values3 = new byte[SIMPLE_PARAM_TABLE_COUNT1];
        for (int i = 0; i < values3.length; ++i)
            values3[i] = (byte)(id + 2);
        final TestBlob testBlob3 = new TestBlob(values3.length, values3);
        row.setBlob3(testBlob3);

        return row;
    }

    private static void checkSimpleParamTableRows(List<SimpleParamTableRow> rows1,
            List<SimpleParamTableRow> rows2)
    {
        assertEquals(rows1.size(), rows2.size());
        for (int i = 0; i < rows1.size(); ++i)
            checkSimpleParamTableRow(rows1.get(i), rows2.get(i));
    }

    private static void checkSimpleParamTableRow(SimpleParamTableRow row1, SimpleParamTableRow row2)
    {
        assertEquals(row1.getId(), row2.getId());
        assertEquals(row1.getName(), row2.getName());
        assertEquals(row1.getBlob1(), row2.getBlob1());
        assertEquals(row1.getBlob2(), row2.getBlob2());
        assertEquals(row1.getBlob3(), row2.getBlob3());

        // check reused explicit count1 parameter
        assertEquals(row2.getBlob1().getCount(), row2.getBlob3().getCount());
    }

    private static final int NUM_SIMPLE_PARAM_TABLE_ROWS = 5;
    private static final int SIMPLE_PARAM_TABLE_COUNT1 = 10;
    private static final int SIMPLE_PARAM_TABLE_COUNT2 = 11;
    private static final String FILE_NAME = "explicit_simple_param_test.sqlite";

    private final File file = new File(FILE_NAME);
    private ExplicitParametersDb database = null;
}

File: test/extensions/language/explicit_parameters/python/ExplicitSimpleParamTest.py
import os

import ExplicitParameters


class ExplicitSimpleParamTest(ExplicitParameters.TestCaseWithDb):
    @classmethod
    def setUpClass(cls):
        super(ExplicitSimpleParamTest, cls).setUpClass()

        test_api = cls.api.explicit_simple_param

        class SimpleParamTableParameterProvider(test_api.SimpleParamTable.IParameterProvider):
            @staticmethod
            def count1(_row):
                return SIMPLE_PARAM_TABLE_COUNT1

            @staticmethod
            def count2(_row):
                return SIMPLE_PARAM_TABLE_COUNT2

        cls.SimpleParamTableParameterProvider = SimpleParamTableParameterProvider

    def setUp(self):
        if os.path.exists(self.dbFileName):
            os.remove(self.dbFileName)
        self._database = self.api.ExplicitParametersDb.from_file(self.dbFileName)
        self._database.create_schema()

    def tearDown(self):
        self._database.close()

    def testDeleteTable(self):
        self.assertTrue(self._isTableInDb())

        testTable = self._database.simple_param_table
        testTable.delete_table()
        self.assertFalse(self._isTableInDb())

        testTable.create_table()
        self.assertTrue(self._isTableInDb())

    def testReadWithoutCondition(self):
        testTable = self._database.simple_param_table

        writtenRows = self._createSimpleParamTableRows()
        testTable.write(writtenRows)

        parameterProvider = self.SimpleParamTableParameterProvider()
        readRows = testTable.read(parameterProvider)
        numReadRows = 0
        for readRow in readRows:
            self._checkSimpleParamTableRow(writtenRows[numReadRows], readRow)
            numReadRows += 1
        self.assertTrue(len(writtenRows), numReadRows)

    def testReadWithCondition(self):
        testTable = self._database.simple_param_table

        writtenRows = self._createSimpleParamTableRows()
        testTable.write(writtenRows)

        parameterProvider = self.SimpleParamTableParameterProvider()
        condition = "name='Name1'"
        readRows = testTable.read(parameterProvider, condition)
        expectedRowNum = 1
        for readRow in readRows:
            self._checkSimpleParamTableRow(writtenRows[expectedRowNum], readRow)

    def testUpdate(self):
        testTable = self._database.simple_param_table

        writtenRows = self._createSimpleParamTableRows()
        testTable.write(writtenRows)

        updateRowId = 3
        updateRow = self._createSimpleParamTableRow(updateRowId, "UpdatedName")
        updateCondition = "id=" + str(updateRowId)
        testTable.update(updateRow, updateCondition)

        parameterProvider = self.SimpleParamTableParameterProvider()
        readRows = testTable.read(parameterProvider, updateCondition)
        for readRow in readRows:
            self._checkSimpleParamTableRow(updateRow, readRow)

    def _createSimpleParamTableRows(self):
        rows = []
        for rowId in range(NUM_SIMPLE_PARAM_TABLE_ROWS):
            rows.append(self._createSimpleParamTableRow(rowId, "Name" + str(rowId)))

        return rows

    def _createSimpleParamTableRow(self, rowId, name):
        values1 = [rowId for i in range(SIMPLE_PARAM_TABLE_COUNT1)]
        testBlob1 = self.api.explicit_simple_param.TestBlob(len(values1), values1)

        values2 = [rowId + 1 for i in range(SIMPLE_PARAM_TABLE_COUNT2)]
        testBlob2 = self.api.explicit_simple_param.TestBlob(len(values2), values2)

        values3 = [rowId + 2 for i in range(SIMPLE_PARAM_TABLE_COUNT1)]
        testBlob3 = self.api.explicit_simple_param.TestBlob(len(values3), values3)

        return (rowId, name, testBlob1, testBlob2, testBlob3)

    def _checkSimpleParamTableRow(self, row1, row2):
        self.assertEqual(row1, row2)

        # check reused explicit count1 parameter
        blob1 = row2[2]
        blob3 = row2[4]
        self.assertEqual(blob1.count, blob3.count)

    def _isTableInDb(self):
        # check if database does contain table
        sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + TABLE_NAME + "'"
        for row in self._database.connection.cursor().execute(sqlQuery):
            if len(row) == 1 and row[0] == TABLE_NAME:
                return True

        return False


TABLE_NAME = "simpleParamTable"
NUM_SIMPLE_PARAM_TABLE_ROWS = 5
SIMPLE_PARAM_TABLE_COUNT1 = 10
SIMPLE_PARAM_TABLE_COUNT2 = 11
